src/coreclr/jit/lsraarm64.cpp

```diff
@@ -1385,9 +1385,104 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
 
     const HWIntrinsic intrin(intrinsicTree);
 
-    int       srcCount      = 0;
-    int       dstCount      = 0;
-    regMaskTP dstCandidates = RBM_NONE;
+    int srcCount = 0;
+
+    // Determine whether this is an operation where an op must be marked delayFree so that it
+    // is not allocated the same register as the target.
+    GenTree* delayFreeOp = getDelayFreeOperand(intrinsicTree);
+
+    // Determine whether this is an operation where one of the ops is an address
+    GenTree* addrOp = LinearScan::getVectorAddrOperand(intrinsicTree);
+
+    // Determine whether this is an operation where one of the ops has consecutive registers
+    bool     destIsConsecutive = false;
+    GenTree* consecutiveOp     = getConsecutiveRegistersOperand(intrinsicTree, &destIsConsecutive);
+
+    // Determine whether this is an operation where one of the ops is an embedded mask
+    GenTreeHWIntrinsic* embeddedOp = getEmbeddedMaskOperand(intrin);
+
+    // If there is an embedded op, then determine if that has an op that must be marked delayFree
+    if (embeddedOp != nullptr)
+    {
+        assert(delayFreeOp == nullptr);
+        delayFreeOp = getDelayFreeOperand(embeddedOp);
+    }
+
+    // Build any immediates
+    BuildHWIntrinsicImmediate(intrinsicTree, intrin);
+
+    // Build all Operands
+    for (size_t opNum = 1; opNum <= intrin.numOperands; opNum++)
+    {
+        GenTree* operand = intrinsicTree->Op(opNum);
+
+        assert(operand != nullptr);
+
+        SingleTypeRegSet candidates = getOperandCandidates(intrinsicTree, intrin, opNum);
+
+        if (embeddedOp == operand)
+        {
+            assert(addrOp != operand);
+            assert(consecutiveOp != operand);
+            assert(delayFreeOp != operand);
+
+            srcCount += BuildEmbeddedOperandUses(embeddedOp, delayFreeOp);
+        }
+        else if (addrOp == operand)
+        {
+            assert(consecutiveOp != operand);
+            assert(delayFreeOp != operand);
+
+            srcCount += BuildAddrUses(operand, candidates);
+        }
+        else if (consecutiveOp == operand)
+        {
+            assert(candidates == RBM_NONE);
+
+            // Some operands have consective op which is also a delay free op
+            srcCount += BuildConsecutiveRegistersForUse(operand, delayFreeOp);
```

Comment:
We also seems to call `buildInternalRegisterUses()` for consecutive registers. Are we missing it here?

Reply:
Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.

- BuildConsecutiveRegistersForUse
- buildInternalRegisterUses
-  BuildDef
- buildInternalRegisterUses
-  BuildDef

Reply:
> We also seems to call `buildInternalRegisterUses()` for consecutive registers. Are we missing it here?

All intrinsics will call `buildInternalRegisterUses()` after the `for` loop, before building the destination.

Reply:
> Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.
> 
> * BuildConsecutiveRegistersForUse
> * buildInternalRegisterUses
> * BuildDef
> * buildInternalRegisterUses
> * BuildDef

In old code:
* op1 - BuildUse (because tgtPrefUse is set, which is because isRMW)
* op2 - BuildConsecutiveRegistersForUse
* op3 - BuildDelayFreeUses
* buildInternalRegisterUses
* BuildDef

In new code:

- delay free = op1 (because isRMW)
- addr = nullptr
- consecutive = op2
- dest consecutive = false
- embedded = nullptr

* BuildHWIntrinsicImmediate (which is a nop)
* op1 - BuildUse (because delayFreeOp == op1)
* op2 - BuildConsecutiveRegistersForUse (because consecutive == op2)
* op3 - BuildDelayFreeUses (because delay free != nullptr)
* buildInternalRegisterUses
* BuildDef

Reply:
You're missing the thing I always miss....

> Also for `NI_AdvSimd_VectorTableLookupExtension()`, we call like this. We should double check that logic in new code.
> 
> * BuildConsecutiveRegistersForUse
> * buildInternalRegisterUses
> * BuildDef

there is a `return srcCount;` here (line 1934)

Which means it never does:

> * buildInternalRegisterUses
> * BuildDef